src/frontend/src/repositories/identityMetadata.ts

```diff
@@ -0,0 +1,195 @@
+
+export type IdentityMetadata = {
+  recoveryPageShownTimestampMillis?: number;
+};
+
+export const RECOVERY_PAGE_SHOW_TIMESTAMP_MILLIS =
+  "recoveryPageShownTimestampMillis";
+
+const convertMetadata = (rawMetadata: MetadataMapV2): IdentityMetadata => {
+  const recoveryPageEntry = rawMetadata.find(
+    ([key]) => key === RECOVERY_PAGE_SHOW_TIMESTAMP_MILLIS
+  );
+  if (recoveryPageEntry === undefined) {
+    return {};
+  }
+  const stringValue = recoveryPageEntry[1];
+  if ("String" in stringValue) {
+    const recoveryPageShownTimestampMillis = Number(stringValue.String);
+    if (isNaN(recoveryPageShownTimestampMillis)) {
+      return {};
+    }
+    return {
+      recoveryPageShownTimestampMillis,
+    };
+  }
+  return {};
+};
+
+type MetadataGetter = () => Promise<MetadataMapV2>;
+type MetadataSetter = (metadata: MetadataMapV2) => Promise<void>;
+type RawMetadataState = MetadataMapV2 | "loading" | "error" | "not-loaded";
+
+/**
+ * Class to manage the metadata of the identity and interact with the canister.
+ *
+ * We decided to not throw any errors because this is non-critical for the application
+ * and we don't want to disrupt user flows if there is an error with the metadata.
+ *
+ * This class loads the metadata in the background when it's created.
+ * It can then be read and updated in memory.
+ * The metadata needs to be committed to the canister with the `commitMetadata` method to persist the changes.
+ */
+export class IdentityMetadataRepository {
+  // The nice IdentityMetadata is exposed to the outside world, while the raw metadata is kept private.
+  // We keep all the raw data to maintain other metadata fields.
+  private rawMetadata: RawMetadataState;
+  // Flag to keep track whether we need to commit the metadata to the canister.
+  private updatedMetadata: boolean;
+  private readonly getter: MetadataGetter;
+  private readonly setter: MetadataSetter;
+
+  static init = ({
+    getter,
+    setter,
+  }: {
+    getter: MetadataGetter;
+    setter: MetadataSetter;
+  }) => {
+    const instance = new IdentityMetadataRepository({ getter, setter });
+    // Load the metadata in the background.
+    void instance.loadMetadata();
+    return instance;
+  };
+
+  private constructor({
+    getter,
+    setter,
+  }: {
+    getter: MetadataGetter;
+    setter: MetadataSetter;
+  }) {
+    this.getter = getter;
+    this.setter = setter;
+    this.rawMetadata = "not-loaded";
+    this.updatedMetadata = false;
+  }
+
+  /**
+   * Load the metadata in the instance variable `rawMetadata`.
+   *
+   * This method won't throw an error if the metadata can't be loaded.
+   * Instead, it will set the instance variable `rawMetadata` to "error".
+   *
+   * @returns {Promise<void>} In case a client wants to wait for the metadata to be loaded.
+   */
+  loadMetadata = async (): Promise<void> => {
+    this.rawMetadata = "loading";
+    try {
+      this.updatedMetadata = false;
+      this.rawMetadata = await this.getter();
```

Comment:
You could also just put the `Promise` as a field. That way you could await it alongside a timer with `Promise.race(...)` in `getMetadata`, no?

Not sure about the error handling though.

Reply:
Yes, if I put the promise there I lose the states and I don't see the advantage. Unless I don't put the promise of the getter directly, but I wrap the getter so that it returns a promise of the states.

What do you think it's the advantage? I can do it in another PR. We would also lose the possibility of a client waiting on `loadMetadata`.